indexer/src/sources/oura_source.rs

```diff
@@ -0,0 +1,215 @@
+    expected_rollback: Option<PointArg>,
+}
+
+impl OuraSource {
+    pub fn new(config: SourceConfig, start_from: Vec<Point>) -> anyhow::Result<Self> {
+        match config {
+            SourceConfig::Oura {
+                network,
+                socket,
+                bearer,
+            } => {
+                let (intersect, rollback) = match start_from {
+                    points if points.is_empty() => {
+                        // we need a special intersection type when bootstrapping from genesis
+                        (IntersectArg::Origin, None)
+                    }
+                    points => {
+                        let (slot_nb, hash) = match points.last().unwrap() {
+                            Point::Origin => {
+                                return Err(anyhow!("Origin point is not supported here"));
+                            }
+                            Point::BlockHeader { slot_nb, hash } => (slot_nb, hash),
+                        };
+                        tracing::info!("Starting sync at block #{} ({})", slot_nb, hash,);
+                        // if last block sync'd was at slot 0,
+                        // that means it was the genesis block so we start from origin
+                        match slot_nb {
+                            0 => (IntersectArg::Origin, None),
+                            _ => {
+                                let point_args: Vec<PointArg> = points
+                                    .iter()
+                                    .flat_map(|p| match p {
+                                        Point::Origin => vec![],
+                                        Point::BlockHeader { slot_nb, hash } => {
+                                            vec![PointArg(*slot_nb, hash.clone())]
+                                        }
+                                    })
+                                    .collect();
+                                let rollback = point_args.first().cloned();
+                                (IntersectArg::Fallbacks(point_args), rollback)
+                            }
+                        }
+                    }
+                };
+
+                let (handles, input) = oura_bootstrap(bearer, intersect, &network, socket)?;
+
+                Ok(OuraSource {
+                    handles,
+                    input,
+                    expected_rollback: rollback,
+                })
+            }
+            _ => Err(anyhow!(
+                "Config {:?} is not supported as oura config",
+                config
+            )),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl dcspark_blockchain_source::Source for OuraSource {
+    type Event = CardanoEventType;
+    type From = Point;
+
+    /// note: from is ignored here since oura is set up just once
+    async fn pull(&mut self, _from: &Self::From) -> anyhow::Result<Option<Self::Event>> {
+        let input = self
+            .input
+            .recv()
```

Comment:
AFAIK this is a blocking function, which doesn't play nice with the async interface. Maybe we should consider using `tokio::task::spawn_blocking` for this? Or a helper thread with an async channel...